### PR TITLE
Tau es

### PR DESCRIPTION
**Pull request by swozniewski**

### CODE REVIEW

File: code_generation/producer.py [hunk 1]
```diff
@@ -68,9 +68,23 @@ def shift(self, name, scope="global"):
...
             raise Exception
         for entry in self.output:
-            entry.shift(
-                name, scope
-            )  # crashes on purpose if output is None. This method should not be called for a producer without output
+            entry.shift(name, scope)
+
+    def ignore_shift(self, name, scope="global"):
+        if not scope in self.scopes:
+            print(
```
Comment (harrypuuter) [inline code_generation/producer.py:75]: use logging instead of print

File: code_generation/producer.py [hunk 2]
```diff
@@ -68,9 +68,23 @@ def shift(self, name, scope="global"):
...
+                "Trying to add shift %s to producer %s in scope %s, but producer does not exist in given scope!"
+                % (name, self.name, scope)
+            )
+            raise Exception
+        if self.output is None:
+            print(
```
Comment (harrypuuter) [inline code_generation/producer.py:81]: logging

File: code_generation/quantity.py [hunk 3]
```diff
@@ -33,20 +34,30 @@ def get_leafs_of_scope(self, scope):
         return [self.name] + [self.name + shift for shift in self.get_shifts(scope)]
 
     def shift(self, name, scope):
+        if scope in self.ignored_shifts.keys():
+            if name in self.ignored_shifts[scope]:
+                log.debug("Ignoring shift {} for quantity {}".format(name, self.name))
+                return
```
Comment (harrypuuter) [inline code_generation/quantity.py:40]: This is a bit inconsistent, in this case the function returns a None, and in all other cases nothing is returned. Either all possible cases should return something, or no case should return something. So i would suggest to remove the return and use an else for the other cases

**Review by harrypuuter — CHANGES_REQUESTED** [review]

File: code_generation/quantity.py (hunk 3 above)
Comment (swozniewski) [reply code_generation/quantity.py:40]: I think a python function without return statement returns None as well.

**Review by harrypuuter — APPROVED** [review]

**Pull request merged**